covid_graphs/covid_graphs/heat_map.py

```diff
@@ -13,7 +13,14 @@
 
 class HeatMap():
     def __init__(self, simulation_pb2_file: Path):
-        reports = create_simulation_reports(simulation_pb2_file)
+        try:
+            reports = create_simulation_reports(simulation_pb2_file)
+            self.parsed_reports = True
+        except:
```

Comment:
I would discourage using a `try` block here as it can hide other exceptions. Non-existence of a file is now a valid state, and can be easily check programatically:
```
if not simulation_pb2_file.is_file():
   ...
   return
```

I also think that from the design point of view this is not the best place for the check. One other possible way to achieve the same would be to move the check to the caller, e.g. along these lines:

- Decouple graph creation from dash dashboard creation
- `HeatMap` initialiser accepts `SimulationResults` proto class
- Dashboard creation reads the proto and instantiates `HeatMap`. If it can't then serves an error page instead of the dashboard.

This way `HeatMap` does not have an ad hoc way of signalling that something went wrong.

I'm not sure if you want to spend effort figuring it out, I'll approve if you don't care that much.

Reply:
I have realised you don't really need approval to merge 😄

Reply:
This code is correct with respect to the current implementation, as `create_simulation_reports` can only fail when the proto is missing or fails to parse.

I agree that that's not the best practice, since code eventually changes. However, I think the right place to parse the proto is inside `simulation_report.py`. Right now no other source file knows about the existence of the `SimulationResults` proto and I think that's great! The only thing that anyone else should know about the proto file is it's path.

What I could agree with is initializing HeatMap with `SimulationReport`, but that means that `create_simulation_reports` might fail (as it's the case now).

I'm really tired now, so I'll think about the right course of action tomorrow..

Reply:
This ended up being super complicate, since HeatMap grew to be a big spaghetti mess. This is the diff of the last commit in `server.py`:
```
-    covid19_heatmap_polynomial_app = HeatMap(simulated_polynomial).create_app(server)
-    covid19_heatmap_exponential_app = HeatMap(simulated_exponential).create_app(server)
+    covid19_heatmap_polynomial_app = create_heat_map_dashboard(simulated_polynomial, server)
+    covid19_heatmap_exponential_app = create_heat_map_dashboard(simulated_exponential, server)
```

I've added a dataclass and a few associated functions. Makes for better error reporting. `create_heat_map_dashboard` might fail if the proto fails to parse.

Reply:
You are right, parsing the proto inside `simulation_report.py` makes sense :) I did not really think through the "perfect" error catching logic, and in the end I think what you had was good enough.